### DISCUSSION TITLE
Programmatically Creating New Tags

### DISCUSSION THREAD

**Discussion by QC8086** [jonaswinkler/paperless-ng#1146] [Support] [answered]
Hello. I am very excited to be using Paperless as it seems like a very convenient system! However, I'm currently attempting to migrate our documents from our (now-unusable) NEAT system. I want to do this programmatically as I have all of the fields I need exported as a CSV and there are some 30,000+ documents. Just dumping them into the consumer is unfortunately not an option as the consumer can't extract a lot of this manually-input-over-years information.

I have most of the details worked out for exporting old documents and importing them with names and dates and such with the REST API, however I need to be able to create tags programmatically. I've been looking through the documentation, and can't find any way to do this. 

It doesn't necessarily have to be through the REST API either. I have looked at the pre-consumption script, and custom parsers pages in the documentation and can't really tell if either of them are capable of what I need.

TL;DR: Basically I just need some way to tell Paperless what documents get what tags (which likely aren't already created), and the name, date, etc. with a programming/scripting language. I am willing to be very flexible with any suggestions or ways this is possible.

Thank you for any help you are willing to provide!

**Comment by Gummientchen** [comment, marked as answer]
You can make a POST request to the endpoint /api/tags/.

If you try to navigate to the url https://YOURPAPERLESS/api/tags/ you can inspect how the request to create a new tag works (at the end of the page).

Example:
`{
    "name": "testTag",
    "colour": null,
    "match": "",
    "matching_algorithm": null,
    "is_insensitive": false,
    "is_inbox_tag": false
}`

**Reply by QC8086** [reply]
Thank you! I had thought I read you could simply "browse" the API at your own paperless instance. However, when I try to go /api (or /api/tags) I get a 404 from Paperless. Do you know if this is something I need to enable, or have I done something wrong with my installation?
In any case, thank you for the help.

**Reply by Gummientchen** [reply]
You have to add a / at the end of the URL or it won't work -> /api/tags/

Took me also some time to figure that out.

**Reply by QC8086** [reply]
You are awesome! Thank you.

**Reply by Gummientchen** [reply]
No problem. Happy importing.
